### ISSUE TITLE
Use OpenTracing to gather valuable execution information and metrics

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/ipfs-cluster#287]
This would be really useful for cluster, to get a feel of which code paths are the slowest and when.

@frrist is working on adding OpenTracing to go-ipfs. We can probably take advantage of this because it's a really good thing to have.

See some work around it:

https://github.com/ipfs/go-ipfs/issues/4487
https://github.com/ipfs/go-ipfs-cmds/pull/39
https://github.com/ipfs/go-ipfs/pull/4506

First work could concentrate in gathering traces for a "pin" request for example.

Once that works, we can start adding tracing for other requests and operations (particularly the recurrent ones i.e. syncs, metric gathering etc).

**Comment by frrist** [comment]
@hsanjuan I have created an [issue here](https://github.com/ipfs/notes/issues/277) to provide a general overview of the OpenTracing API (and a few discussion points) - any feedback would be appreciated!

**Comment by s1na** [comment]
@hsanjuan I would be interested in working on this, however the tag suggests this to be a hard task. Do you think I can take a stab? and if so, is it okay if I do it in smaller steps?

**Comment by hsanjuan** [comment]
@s1na It would be amazing if you worked on this. I suggest you start by adding tracing for a small thing (like the `ID` method, and once we have a handle on that we can think of doing more interesting traces like the whole `Pin` operations.

**Comment by hsanjuan** [comment]
@s1na I saw you asked on IRC about how to import some things. You should not need to import anything new in terms of dependencies. According to @frrist, the last version of go-log (as of current master), supports open-tracing out of the box

https://godoc.org/github.com/ipfs/go-log#EventLogger

This is very new and @ffrist was going to post here with some details.

You can also check the PR in question: https://github.com/ipfs/go-log/pull/20/files . Note that godoc is no help in terms of documentation, because the methods which are documented are part of a non-exported type, but the comments in the source code seem useful.

**Comment by s1na** [comment]
@hsanjuan Oh, sorry for my misunderstanding. I thought you meant we should also develop something similar.
So from what I understood, Eventlogger acts as a wrapper around opentracing's api, and there's also a `LoggableTracer` and `LoggableReporter` provided which can be used to write the encoded trace to a chosen `io.writer`, which is mostly used in tests, and in https://github.com/ipfs/notes/issues/277 an ipfs plugin for jaeger tracer is being used. I didn't know if we can also integrate ipfs plugins in ipfs-cluster? I thought we can't, and that's why attempted to use jaeger directly for the tracer.
How should I proceed?

**Comment by hsanjuan** [comment]
Hey,

ok, as far as I understand (sorry, I had to do some reading I've been postponing):

1. the first step is to add tracing. That is, using go-log EventLogger to start and finish spans in the source code.

2. The second step is to add a tracer, for example `ipfs-cluster-service --trace` would create an `opentracer.Tracer` instance (for example simply using https://godoc.org/github.com/ipfs/go-log/tracer). Once created this is set with https://godoc.org/github.com/opentracing/opentracing-go#SetGlobalTracer. And go-log will just send spans to it (and they should show on the screen).

3. Third step is to allow using tracers like jaeger. go-ipfs uses https://github.com/ipfs/go-jaeger-plugin . It is gxed, but the module itself has vendored deps so I would not include it directly. In this light, t makes sense to support plugins in cluster in a similar way as go-ipfs does, so we can re-use things like the ipfs/go-jaeger-plugin.

1 and 2 should be relatively easy (at least to start with). Having EventLogger calls in the right places to produce a trace for, say, all the pinning procedure is probably some work already, regardless of the tracer used. Once we have some tracing like that and the basic loggableTracer, we can think how to best approach the plugin system to support additional, external pieces of code like the jaeger plugin.

Does this make it a bit clearer? Hopefully I'm getting things right.. but let me know if you think I misunderstood something about how the whole tracing thing works.

**Comment by frrist** [comment]
Hey @s1na - thank you for picking this up!
>from what I understood, Eventlogger acts as a wrapper around opentracing's api

Correct.

>there's also a LoggableTracer and LoggableReporter provided which can be used to write the encoded trace to a chosen io.writer

Correct, although something to note here. The `LoggableTracer` does not record _all_ data about a span. It's output does not contain, `TraceID`, `SpanID` or `SpanContext`. 
I have linked to examples on how this works with go-ipfs [here](https://github.com/ipfs/go-log#tracing) - basically it's the writer for the `ipfs log tail` command.

>How should I proceed?

I would start by importing jaeger, and setting it as the global tracer -`opentracing.SetGlobalTracer(jaegerTracer)`. Then create some spans using the `Start` and `Finish` methods from `go-log`.
Ideally we will be able to reuse [go-jaeger-plugin](https://github.com/ipfs/go-jaeger-plugin), but I am unsure if `ipfs-cluster` supports plugins at this time.

**Comment by hsanjuan** [comment]
Hey @s1na , I think you missed this comment on IRC:

```
00:27:45 <frist> Hi s1na, we added opentracing to go-log because go-log is widely used by projects in ipfs land and would allow code that was already instrumented with Event logging to produce tracing information without requiring changes to each package (other than updating the gx version). There were definitely some usability trade-offs made there, one which you already pointer out - an HttpCarrier, or lack thereof.
00:28:05 We only have support for binary carrier at the moment because that was all that was necessary for go-ipfs (I apologize for not considering a larger project scope, like ipfs-cluster) to pass SpanContext between nodes. For example in the case of DHT methods, the spancontext can be serialized to bytes and added to a protobuf message as an optional field. We would definitely be open to the addition of and
00:28:06 HTTP carrier in go-log if that helps ipfs-cluster tracing.
00:29:36 Another thing to note, the tracer in go-log is only for writing traces as event logs (meaning it strips out things like TraceID, and SpanContext). This was done so that `ipfs log tail` could be used to view the original event logs and the traces generated by opentracing in the terminal as if they were event logs.
```
Also, I add your own comments for context:

```
12:34:08 <s1na> hsanjuan: to be honest, i think the simplest way would be to use opentracing directly
12:36:21 it's a well-defined standard and quite flexible and easy to use by itself. and morever it seems it me that from a logical pov, logging and tracing are different things. thats why i asked the second question
12:40:32 but as you say, there will be time to iterate. i will then define separate traces for now. thanks for the help
```

> i think the simplest way would be to use opentracing directly

By now you've probably spent more time playing around with it than I have looked at the docs. Your reasoning makes sense and cluster is not instrumented with event logging, but also it seemed easy to re-use the wrapping provided by go-log. **I trust your judgement here and I'm going to leave the decision up to you**. It would be nice that when I do `--trace`, some legible tracing appears in the logs though.

**Comment by hsanjuan** [comment]
@s1na am I correct suspecting that having successful tracing will require that all the methods we want to trace receive a ctx, as that ctx is wrapping the SpanContexts that make the trace?

So, when the HTTP API component receives a call (`/id`), let's assume:

* it should create a span from it's root context (`StartSpanFromContext`, let's assume no http carrier yet)
* then it will make an RPC call passing the context that it received from StartSpanFromContext,
* then the RPC call will call a method in the RPC server (`ClusterID`) with a context (that should be wrapping the span), so that it can again derivate a new span and new context with `StartSpanFromContext`, 
* and then that method should call `Cluster.ID` with that context so that it can yet make another span from the parent one.

This would require:

* Modifying the signature of `Cluster.ID` to take a context (or adding `Cluster.idWithCtx` auxiliary function)
* Having gorpc support tracing (with a carrier) so that the context received by the method in the server wraps the SpanContext from the RPC client. Note this already works for local rpc calls since the context is passed all the way.

**Comment by s1na** [comment]
@hsanjuan @frrist Sorry if this is not the best place to discuss this, but this just occured to me today and has kept my thought busy since, I thought I'd ask your opinions on it:

Assumption: Distributed tracing is most useful in production (which is possible due to lower overhead, in comparison to profiling, although [google does profiling also in prod](https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b)), and aggregated data yields information that can't be inferred from single traces.

Observation: Most services which use tracing run on clusters, with traces propagating from one microservice to the other throughout the lifetime of a request (load balancer -> auth -> db ->...). Traces are collected, sampled and sent to a tracing server.

Challenge: Decentralized applications such as ipfs are made up of peers, and do not run in a private cluster, do not have a (scalable) central server that can act as a tracing backend for the whole network. How do we sample and aggregate data from each peer in a real running network, and where do we report and store it for usage? Is tracing as useful for p2p systems anyway? (I think so) Can there be privacy issues with tracing data? Is it possible to write a tracing system on top of ipfs?

**Comment by frrist** [comment]
> am I correct suspecting that having successful tracing will require that all the methods we want to trace receive a ctx, as that ctx is wrapping the SpanContexts that make the trace?

Correct. This is how opentracing manages spans.
- ContextWithSpan
https://github.com/opentracing/opentracing-go/blob/8013ba5856cd701278d1111c5b4dd69e580a86c9/gocontext.go#L9-L13
- SpanFromContext
https://github.com/opentracing/opentracing-go/blob/8013ba5856cd701278d1111c5b4dd69e580a86c9/gocontext.go#L15-L27

>  How do we sample and aggregate data from each peer in a real running network, and where do we report and store it for usage?

This is a great question! I have only collected tracing data from IPFS nodes that I run locally using [IPTB](https://github.com/ipfs/iptb). A couple ideas for this to work in a real network, or maybe a testnet. 
- Require participating nodes to connect to the same tracer (not really scaleable). 
- Nodes store their tracing data (serialized SpanContext) in a publicly accessible area so that it could be scraped later for post processing (not sure if any tracer implementations currently support post processing)

This is definitely one of the hardest problems to solve wrt to tracing in a peer to peer system.

**Comment by hsanjuan** [comment]
Fixed by #486
